### PR TITLE
Fix stack handling in POWER SIGTRAP handler (#12482)

### PR DESCRIPTION
**Pull request by ghost**
As mentioned during the incomplete analysis of issue #12482 [here](https://github.com/ocaml/ocaml/issues/12482#issuecomment-1710194186), the `SIGTRAP` handler logic is applied to an incorrect stack pointer value.

This PR implements the suggestion made in the second-to-last paragraph of the comment linked above, to give the signal handler enough information to compute proper data.

Also linking to #12276 where this code was introduced.

### CODE REVIEW

**Comment by ghost** [comment]
(note that I'll be AFK from this evening until september 20th, so there is no hurry to review this and I won't be able to address any comments until then)

File: runtime/signals_nat.c [hunk 1]
```diff
@@ -106,6 +106,10 @@ void caml_sigtrap_handler(int signo, siginfo_t * info, void * context)
   uint64_t ctx_sp = ctx->uc_mcontext.gp_regs[1];
   uint64_t ctx_exn_ptr = ctx->uc_mcontext.gp_regs[29];
   uint64_t ctx_young_ptr = ctx->uc_mcontext.gp_regs[31];
+  uint64_t ctx_r11 = ctx->uc_mcontext.gp_regs[11];
+  /* r11 has been set to the offset between the current stack pointer and
+     the caller's frame. */
+  ctx_sp += ctx_r11;
```
Comment (xavierleroy) [inline runtime/signals_nat.c]: This doesn't feel right.  The sp value that is stored in `dom_st->current_stack-> sp` a few lines below should be the "true" sp, not this adjusted sp!  The adjusted sp is only for saving `ctx_pc`.

**Review by xavierleroy — COMMENTED** [review]
Something looks strange to me, see below.

At any rate, the main motivation to stick with the trap-based implementation of bounds checks (besides the pleasure of using the rare `tdlle` instruction) was code compactness: the check is done in 1 instruction instead of 2 for the usual compare-and-branch approach.  With this PR, we're back to 2 instructions for the trap-based approach, and it's getting hard to justify it.

File: runtime/signals_nat.c (hunk 1 above)
Comment (ghost) [reply runtime/signals_nat.c]: Fixed. Does not change the outcome of the test though.

**Pull request closed without merging**